ee/tools/puppet/fleetdm/lib/puppet/functions/fleetdm/preassign_profile.rb

```diff
@@ -12,28 +12,59 @@
+    # initialize our response
+    preassign_profile_response = { 'error' => '', 'resource_changed' => false }
+
     client = Puppet::Util::FleetClient.instance
     env = closure_scope['server_facts']['environment']
     run_identifier = "#{closure_scope.catalog.catalog_uuid}-#{Puppet[:node_name_value]}"
-    response = client.preassign_profile(run_identifier, host_uuid, template, group, ensure_profile, closure_scope['environment'])
 
-    if response['error'].empty?
-      base64_checksum = Digest::MD5.base64digest(template)
-      host = client.get_host_by_identifier(host_uuid, env)
-      host_profiles = client.get_host_profiles(host['body']['host']['id'], env)
+    # initiate the pre-assignment process with fleet server
+    client_resp = client.preassign_profile(run_identifier, host_uuid, template, group, ensure_profile, closure_scope['environment'])
+    unless client_resp && client_resp['error']&.empty?
+      Puppet.err("Error pre-assigning profile #{profile_identifier} (ensure #{ensure_profile}): #{client_resp&.[]('error')} \n\n #{template}")
+      preassign_profile_response['error'] = client_resp['error']
+      return preassign_profile_response
+    end
+
+    # get host by idenfifier to get the host id
+    client_resp = client.get_host_by_identifier(host_uuid, env)
+    unless client_resp && client_resp['error']&.empty?
+      Puppet.err("Error getting host by identifier #{host_uuid}: #{client_resp&.[]('error')} \n\n #{template}")
+      preassign_profile_response['error'] = client_resp['error']
+      return preassign_profile_response
+    end
+    unless client_resp['body'] && client_resp['body']['host'] && client_resp['body']['host']['id']
+      Puppet.err("No host found for #{host_uuid} \n\n #{template}")
+      preassign_profile_response['error'] = client_resp['error']
+      return preassign_profile_response
+    end
+
+    # get host profiles currently assigned to the host
+    client_resp = client.get_host_profiles(client_resp['body']['host']['id'], env)
+    unless client_resp && client_resp['error']&.empty?
+      Puppet.err("Error getting host profiles for #{host_uuid}: #{client_resp&.[]('error')} \n\n #{template}")
+      preassign_profile_response['error'] = client_resp['error']
+      return preassign_profile_response
+    end
 
-      if host_profiles['error'].empty?
-        Puppet.info("successfully pre-set profile #{profile_identifier} as #{ensure_profile}")
+    # if this is the first run on the device, profiles will be empty so we can skip the checksum
+    # comparison and mark the resource as changed depending on the ensure_profile value
+    unless client_resp['body'] && client_resp['body']['profiles'] && !client_resp['body']['profiles']&.empty?
+      Puppet.info("No assigned profiles found, this may be the first run for #{host_uuid}")
+      preassign_profile_response['resource_changed'] = ensure_profile == 'present'
+      return preassign_profile_response
+    end
```

Comment:
this is just an idea up for discussion. I personally find the actual logic a bit hard to follow with all the error handling/checking.

what do you think about doing something like this? https://github.com/fleetdm/fleet/pull/16121/files

Reply:
I went with your approach with some additional tweaks. Lemme know what you think :)